### PR TITLE
Avoid temp val in property functions when the arguments are val already.

### PR DESCRIPTION
**Pull request by xbcnn**
In many cases, we use fixed keys to set/get properties of val object, and
the keys can be reused.

This helps to avoid temporary key val objects being created and no bother
to refcount when pass val directly.

### CODE REVIEW

**Comment by kripken** [comment]
Thanks for the PR @xbcnn !

@RReverser do you know this, and have time perhaps? No worries if not!

**Comment by RReverser** [comment]
I don't see why allowing raw handles is useful / necessary in this context tbh. You can already reuse fixed keys by storing them as `val` instances.

Raw `EM_VAL` handle is meant to be used only for interop with JS snippets rather than in normal C++ code.

**Comment by xbcnn** [comment]
Pass val will call into JS to increase refcount, which is I want to avoid mainly here.

Also a temporary val created with the same handle. The JS object's refcount will bounce with the temp val.  In my using cases, I think they are unnecessary and not good for performance.

**Comment by RReverser** [comment]
I wouldn't expect it to be particularly big - as you said, it's just a refcount increase. Tbh Embind isn't very performant solution by design, even though it's often the most convenient one.

If you really see overhead from Embind just crossing JS/C++ boundary in your real-world app, you might instead try using `EM_JS` snippets (see https://web.dev/emscripten-embedding-js-snippets/ for more details). They'd allow you to both minimize boundary crossings, and write static JS code that can be optimized by engines.

Whether it's worth the changes to the code, really depends on your perf measurements though.

**Comment by xbcnn** [comment]
Thanks. 
The EM_JS solution is not suitable for me. And I do have cases which reuses the key val very frequently.

> Raw `EM_VAL` handle is meant to be used only for interop with JS snippets rather than in normal C++ code.
I understand your concern about using the raw 'EM_VAL' handle other than interop with JS.  But I also think when pass val as key for property access, a temporary val, and with it adding/releasing refcount are unnecessary costs and unwise.

How about this: partial specialization the key type, when it's emscripten::val, it's obvious that it wants to reuse the val (and JS object behind), we just pass its raw handle instead a copied temp and incur refcount bounce.

**Comment by RReverser** [comment]
> How about this: partial specialization the key type, when it's emscripten::val, it's obvious that it wants to reuse the val (and JS object behind), we just pass its raw handle instead a copied temp and incur refcount bounce.

I think such specialization is reasonable, yeah. However, instead of specializing each function that does `val(something).handle`, I suggest extracting an internal helper `get_val_handle()` and specializing it for `val&` and `T&` separately.

Then you can replace any existing usages of `val(something).handle` with that helper which would make sure that all of them take the faster route when possible.

**Comment by RReverser** [comment]
> The EM_JS solution is not suitable for me. And I do have cases which reuses the key val very frequently.

I'd be curious to hear more about this btw. Why doesn't it work for you?

**Comment by xbcnn** [comment]
> I suggest extracting an internal helper `get_val_handle()` and specializing it for `val&` and `T&` separately.
> 
> Then you can replace any existing usages of `val(something).handle` with that helper which would make sure that all of them take the faster route when possible.

Nice suggestion!  While `get_val_handle()` seems not work if we pass T& and create temp val inside, as the returned `EM_VAL` handle becomes invalid immediately.  i.e.
          EM_VAL get_val_handle(const T& v) {  return val(v).handle; }
We need keep the temp val in the using place.  I'd like to use `wrapper_val()` which return val instead.

**Comment by xbcnn** [comment]
> > The EM_JS solution is not suitable for me. And I do have cases which reuses the key val very frequently.
> 
> I'd be curious to hear more about this btw. Why doesn't it work for you?

It's something like a bound function which is called by JS side, and returns a val with computed results.  The val is composed of a set of fixed keys, while values will change.

**Comment by RReverser** [comment]
> We need keep the temp val in the using place. I'd like to use `wrapper_val()` which returns val instead.

Ah, you're right, of course. I'm happy to take a look at other solutions.

File: system/include/emscripten/val.h [hunk 1]
```diff
@@ -624,6 +610,15 @@ namespace emscripten {
                     argv));
         }
 
+        template<typename T>
+        val wrapper_val(const T& v) const {
+            return val(v);
+        }
+
+        val wrapper_val(const val& v) const {
+            return v;
```
Comment (RReverser) [inline system/include/emscripten/val.h:619]: This still creates a copy, no?

**Comment by xbcnn** [comment]
> > It's something like a bound function which is called by JS side, and returns a val with computed results. The val is composed of a set of fixed keys, while values will change.
> 
> Okay... I still don't see why that wouldn't work with EM_JS though. As shown in the linked article, you can use handle APIs to do all the work in EM_JS snippet and then convert the result to `val` just once.

My understanding about EM_JS macro is it wraps JS snippet as function for C++, the computations happen in JS side and result back to C++ maybe via val.  My case is just the opposite: JS side initiates invoking, C++ side does the computations and returns a val back to JS.

**Comment by RReverser** [comment]
Right, but what I'm saying is, you could still do computations in C++ but then use EM_JS to just construct a JS object in one go, without intermediate `val`s for different properties.

**Comment by RReverser** [comment]
Not without Embind in direction, no.

Anyway, as for this PR - I think it's looking good, but I'd rename the very generic name `wrapper_val` (who knows what it wraps or how many more wrappers we'll need) into something more descriptive. `val_ref`?

**Comment by xbcnn** [comment]
@RReverser  Is there any further concerns about the PR?

**Comment by RReverser** [comment]
Ah, one more - looks like there's a missed `val_ref` opportunity in `delete_` at least:

https://github.com/emscripten-core/emscripten/blob/12799e987a78bc18b3e2858a773186e8170e01f0/system/include/emscripten/val.h#L578-L580

**Comment by xbcnn** [comment]
> while on vacation (and a bit sick)

Sorry for hearing that, take care!



> Ah, one more - looks like there's a missed `val_ref` opportunity in `delete_` at least:
> 
> https://github.com/emscripten-core/emscripten/blob/12799e987a78bc18b3e2858a773186e8170e01f0/system/include/emscripten/val.h#L578-L580

Oh, really. I didn't noticed it.  I'd like to move it up close to the get/set property functions.

**Comment by RReverser** [comment]
> Oh, really. I didn't notice it before. I'd like to move it up close to the get/set property functions.

That's fine by me.

**Comment by xbcnn** [comment]
> do computations in C++ but then use EM_JS to just construct a JS object in one go, without intermediate `val`s for different properties.

btw, I tested this method a days ago. I found it is about 5~8 times fast in the tests. Helpful! Thanks. Now I'm trying to apply it to our production codes in several hot paths.

**Comment by RReverser** [comment]
I'm glad :) It does usually often measurable speed-ups, that's why I was trying to promote it.

**Comment by RReverser** [comment]
@kripken @sbc100 Does one of you want to take a look at the PR before merging?

**Comment by kripken** [comment]
Looks like there are test failures - merging in latest main should help.

**Comment by xbcnn** [comment]
> Looks like there are test failures - merging in latest main should help.

@kripken  Do you mean that I should rebase my branch `update_val` ?

**Comment by kripken** [comment]
This should do it:
```
git fetch origin # where origin is emscripten upstream
git merge
git push # to your branch for your PR
```
Basically you need to get the latest commits from main merged into this branch. (Don't worry about merge commits, we'll squash during landing anyhow.)

**Comment by xbcnn** [comment]
@RReverser @kripken  Would you please help to land this?

File: system/include/emscripten/val.h [hunk 2]
```diff
@@ -489,22 +489,22 @@ namespace emscripten {
...
 
         template<typename K, typename V>
         void set(const K& key, const V& value) {
-            internal::_emval_set_property(handle, val(key).handle, val(value).handle);
+            internal::_emval_set_property(handle, val_ref(key).handle, val_ref(value).handle);
+        }
+
+        template<typename T>
+        bool delete_(const T& property) const {
+            return internal::_emval_delete(handle, val_ref(property).handle);
```
Comment (sbc100) [inline system/include/emscripten/val.h:507]: Why move this method up?  It seems to just make the diff harder to read.

**Review by sbc100 — APPROVED** [review]

**Pull request merged**